raster_loader/io/databricks.py

```diff
@@ -0,0 +1,327 @@
+    update_metadata,
+)
+
+from typing import Iterable, List, Tuple
+
+try:
+    from databricks.connect import DatabricksSession
+except ImportError:  # pragma: no cover
+    _has_databricks = False
+else:
+    _has_databricks = True
+
+from raster_loader.io.datawarehouse import DataWarehouseConnection
+
+
+class DatabricksConnection(DataWarehouseConnection):
+    def __init__(self, server_hostname, access_token, cluster_id, parallelism=200):
+        # Validate required parameters
+        if not server_hostname:
+            raise ValueError("server_hostname cannot be null or empty")
+        if not access_token:
+            raise ValueError("access_token cannot be null or empty")
+        if not cluster_id:
+            raise ValueError("cluster_id cannot be null or empty")
+        if not _has_databricks:
+            import_error_databricks()
+
+        # Normalize server_hostname by removing any 'https://' prefix
+        self.server_hostname = server_hostname.replace("https://", "")
+        self.access_token = access_token
+        self.cluster_id = cluster_id
+        self.parallelism = parallelism
+
+        try:
+            self.spark = DatabricksSession.builder.remote(
+                host=f"https://{self.server_hostname}",
+                token=access_token,
+                cluster_id=cluster_id,
+            ).getOrCreate()
+        except Exception as e:
+            print(f"Error initializing Spark session: {e}")
+            raise
+
+    def execute(self, query: str) -> list:
+        """Execute a SQL query and return results as a list of rows"""
+        try:
+            return self.spark.sql(query).collect()
+        except Exception as e:
+            print(f"Error executing query: {e}")
+            raise
+
+    def execute_to_dataframe(self, query: str) -> "pd.DataFrame":
+        """Execute a SQL query and return results as a pandas DataFrame"""
+        try:
+            return self.spark.sql(query).toPandas()
+        except Exception as e:
+            print(f"Error executing query: {e}")
+            raise
+
+    def quote(self, q):
+        if isinstance(q, str):
+            q = q.replace("'", "''")
+            return f"'{q}'"
+        return str(q)
+
+    def quote_name(self, name):
+        """Quote a table name with proper escaping for Databricks."""
+        parts = name.replace("`", "").split(".")
+        return ".".join(f"`{part}`" for part in parts)
+
+    def upload_records(
+        self,
+        records: Iterable,
+        fqn: str,
+        overwrite: bool = False,
+        parallelism: int = 1000,
+    ):
+        records_list = []
+        for record in records:
+            if "metadata" in record:
+                del record["metadata"]
+            records_list.append(record)
+
+        if not records_list:
+            print("No records to upload.")
+            return True
+
+        try:
+            # Convert to Pandas DataFrame
+            data_df = pd.DataFrame(records_list)
```

Comment:
if a dataframe is created here, is it not better and more efficient to filter the metadata in the dataframe instead of doing it in a "for loop" in line 93?